File: blurt_py_80211/crc.py
#!/usr/bin/env python
import numpy as np
import util
from scipy import weave
from scipy.weave import converters

def remainder1(a, b):
    # clobbers input!
    for i in xrange(a.size-b.size+1):
        if a[i]:
            a[i:i+b.size] ^= b
    return a[1-b.size:]

# # remainder1 rewritten in terms of bit arithmetic
# def remainder2(a, b):
#     N = a.size
#     M = b.size-1
#     m = (1<<M)-1
#     L = 8
#     r = sum(a[M-1::-1] << np.arange(M))
#     a = util.shiftin(a[:M-1:-1], L)[::-1]
#     b = sum(b[M:0:-1] << np.arange(M))
#     for aa in a:
#         for l in xrange(L):
#             r = ((r << 1) & m) ^ ((aa >> (L-1-l)) & 1) ^ (b if (r >> (M-1)) & 1 else 0)
#     return (r >> np.arange(M)[::-1]) & 1

# # remainder2 rewritten in terms of a look-up table
# def remainder4(a):
#     if a.size % L:
#         a = np.r_[a[::-1], np.zeros(-a.size%L, int)]
#     else:
#         a = a[::-1]
#     r = 0
#     a = util.shiftin(a, L)[::-1]
#     for i in xrange(0, a.size):
#         r = ((r << L) & m) ^ a[i] ^ _lut[r >> s]
#     return (r >> np.arange(M)[::-1]) & 1

def remainder5(a, no_shift_out=False):
    if a.size % L:
        a = np.r_[a[::-1], np.zeros(-a.size%L, int)]
    else:
        a = a[::-1]
    a = util.shiftin(a, L)[::-1]
    A = a.size
    rf = np.zeros(1, np.uint32)
    code = """
    uint32_t r = 0;
    for (int i=0; i<A; i++)
        r = ((r << L) & m) ^ a(i) ^ _lut(r >> s);
    rf(0) = r;
    """
    weave.inline(code, ['A', 'rf', 'L', 'm', 'a', '_lut', 's'], type_converters=converters.blitz)
    if no_shift_out:
        return rf[0]
    else:
        return (rf[0] >> np.arange(M)[::-1]) & 1

def lut_bootstrap(b, new_L):
    global L, _lut, M, s, M_L, m
    M = b.size-1
    if L is None:
        fn = lambda a: util.shiftin(remainder5(a, b)[::-1], M)[0]
    else:
        fn = lambda a: remainder5(a, True)
    lut = np.empty(1<<new_L, int)
    for i in xrange(lut.size):
        lut[i] = fn(np.r_[util.shiftout(np.r_[i], new_L)[::-1], np.zeros(M, int)])
    _lut = lut
    L = new_L
    s = M-L
    M_L = M/L
    m = (1<<M)-1

def lut_dump(fn):
    import cPickle
    with open(fn, 'wb') as f:
        cPickle.dump((_lut,M,L,s,M_L,m), f)

def lut_load(fn):
    import cPickle
    global _lut, M, L, s, M_L, m
    try:
        with open(fn, 'rb') as f:
            _lut, M, L, s, M_L, m = cPickle.load(f)
        return True
    except:
        return False

def FCS(calculationFields):
    k = calculationFields.size
    return 1 & ~remainder(np.r_[np.ones(32, int), np.zeros(k, int)] ^ np.r_[calculationFields, np.zeros(32, int)], G)

def checkFCS(frame):
    k = frame.size
    x = remainder(np.r_[np.ones(32, int), np.zeros(k, int)] ^ np.r_[frame, np.zeros(32, int)], G)
    return all(x == correct_remainder)

remainder = lambda a, b: remainder5(a)

# The FCS field is transmitted commencing with the coefficient of the highest-order term.
# polynomial from x^32 down to x^0: 0b100000100110000010001110110110111
CRC32_802_11_FCS_G = 0b100000100110000010001110110110111
CRC32_802_11_FCS_remainder = 0b11000111000001001101110101111011
G = util.shiftout(np.r_[CRC32_802_11_FCS_G], 33)[::-1]
correct_remainder = util.shiftout(np.r_[CRC32_802_11_FCS_remainder], 32)[::-1]

L = None
fn = 'crc_lut_16'
if not lut_load(fn):
    lut_bootstrap(G, 8)
    lut_bootstrap(G, 16)
    lut_dump(fn)

File: blurt_cpp_80211/crc.cpp
#include "crc.h"
#include <algorithm>
#include "util.h"
#include <stdint.h>

void CRC::remainder_slow(const bitvector &a, bitvector &output) {
    bitvector c(a);
    int C = c.size(), B = b.size();
    for (int i=0; i<C-B+1; i++)
        if (c[i])
            for (int j=0; j<B; j++)
                c[i+j] = c[i+j] ^ b[j];
    output.assign(c.rbegin(), c.rbegin()+(B-1));
}

uint32_t CRC::remainder_fast(const bitvector &a, bitvector &output) {
    int N = a.size(), A = (N+L-1)/L;
    std::vector<int> a_words(A);
    for (int i=0; i<A; i++) {
        int x = 0, k = N-1-i*L;
        for (int j=0; j<L && j <= k; j++)
            x += a[k-j] << j;
        a_words[A-1-i] = x;
    }
    uint32_t r = 0;
    for (int i=0; i<A; i++)
        r = ((r << L) & m) ^ a_words[i] ^ lut[r >> s];
    output.resize(M);
    for (int i=0; i<M; i++)
        output[i] = (r >> (M-1-i)) & 1;
    return r;
}

void CRC::lut_bootstrap(const bitvector &new_b, int new_L) {
    b.assign(new_b.begin(), new_b.end());
    int new_M = b.size()-1;
    std::vector<uint32_t> new_lut(1<<new_L);
    if (L == 0) {
        // no look-up table; fall back on slow code
        std::vector<int> results;
        for (int i=0; i<1<<new_L; i++) {
            std::vector<int> num(1);
            num[0] = i;
            bitvector a;
            shiftout(num, new_L, a);
            std::reverse(a.begin(), a.end());
            a.resize(a.size() + new_M);
            bitvector rem;
            remainder_slow(a, rem);
            shiftin(rem, new_M, results);
            new_lut[i] = results[0];
        }
    } else {
        bitvector bits;
        for (int i=0; i<1<<new_L; i++) {
            std::vector<int> num(1);
            num[0] = i;
            bitvector a;
            shiftout(num, new_L, a);
            std::reverse(a.begin(), a.end());
            a.resize(a.size() + new_M);
            new_lut[i] = remainder_fast(a, bits);
        }
    }
    lut.assign(new_lut.begin(), new_lut.end());
    M = new_M;
    L = new_L;
    s = M-L;
    m = (1ull<<M)-1;
}

void CRC::FCS(const bitvector &calculationFields, bitvector &output) {
    bitvector a;
    a.assign(calculationFields.begin(), calculationFields.end());
    a.resize(a.size() + 32);
    for (int i=0; i<32; i++)
        a[i] = !a[i];
    remainder_fast(a, output);
    for (int i=0; i<output.size(); i++)
        output[i] = !output[i];
}

bool CRC::checkFCS(const bitvector &frame) {
    bitvector a;
    a.assign(frame.begin(), frame.end());
    a.resize(a.size() + 32);
    for (int i=0; i<32; i++)
        a[i] = !a[i];
    bitvector output;
    remainder_fast(a, output);
    return std::equal(output.begin(), output.end(), correct_remainder.begin());
}

static const bool CRC32_802_11_FCS_G[] = {1,0,0,0,0,0,1,0,0,1,1,0,0,0,0,0,1,0,0,0,1,1,1,0,1,1,0,1,1,0,1,1,1};
static const bool CRC32_802_11_FCS_remainder[] = {1,1,0,0,0,1,1,1,0,0,0,0,0,1,0,0,1,1,0,1,1,1,0,1,0,1,1,1,1,0,1,1};

CRC::CRC() : correct_remainder(CRC32_802_11_FCS_remainder, CRC32_802_11_FCS_remainder+32), L(0) {
    bitvector G(CRC32_802_11_FCS_G, CRC32_802_11_FCS_G+33);
    lut_bootstrap(G, 8);
    lut_bootstrap(G, 16);
}
